### PR TITLE
Make sure code_typet always has parameters property

### LINKED ISSUE #2083: Spurious function call type difference in goto-diff
Some function calls have `ID_parameters`, others don't.

This is observed in failing instruction->code comparisons at:
https://github.com/diffblue/cbmc/blob/develop/src/goto-diff/unified_diff.cpp#L396
https://github.com/diffblue/cbmc/blob/develop/src/goto-diff/syntactic_diff.cpp#L98

The difference is not visible in --show-goto-functions output. It requires dumping ireps.

A temporary fix has been added here: https://github.com/diffblue/cbmc/commit/0b3170d425ed5aa76e2c92c13cecae87098f271e

So far, it seems that only clinit wrappers are affected.

### PR DESCRIPTION
**Pull request by peterschrammel**
The existence of the parameters property may depend on
whether parameters() has been called or not, which can
lead to unexpectedly unequal instructions in goto programs.

Fixes #2083

### CODE REVIEW

**Review by tautschnig — APPROVED** [review]
Just a procedural suggestion: split this in two commits, with the first one doing the changes to std_types.h and the second one being a clean revert of 0b3170d425.

**Comment by kroening** [comment]
I would then do the same with return_type().
Also, I'd mark the () constructor as deprecated, as this is another case of "problem caused by partially-constructed object".

**Review by kroening — CHANGES_REQUESTED** [review]
do same with return_type, and make () constructor deprecated.

**Review by smowton — APPROVED** [review]
As per @tautschnig, once @kroening's reservations are addressed

**Comment by peterschrammel** [comment]
@kroening, I've introduced new constructors, deprecated the old one and cleaned up its occurrences.

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 1]
```diff
@@ -760,15 +760,14 @@ void java_bytecode_convert_classt::add_array_types(symbol_tablet &symbol_table)
...
-      java_reference_type(symbol_typet("java::java.lang.Object"));
     code_typet::parametert this_param;
     this_param.set_identifier(id2string(clone_name)+"::this");
     this_param.set_base_name("this");
     this_param.set_this();
     this_param.type()=java_reference_type(symbol_type);
-    clone_type.parameters().push_back(this_param);
+    code_typet clone_type(
+      {this_param},
+      java_reference_type(symbol_typet("java::java.lang.Object")));
```
Comment (kroening) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: Note that we've got java_lang_object_type()

**Comment by tautschnig** [comment]
Not there isn't enough work in here already, but "Use the two-param code_typet constructor" would now also enable a lot of use of `const` for almost all of the `code_typet` objects being declared.

**Comment by kroening** [comment]
Note that this also tweaks the remove_returns behavior, which may not have been intended.

**Review by kroening — APPROVED** [review]

**Comment by tautschnig** [comment]
> Note that this also tweaks the remove_returns behavior, which may not have been intended.

In what way is that impacted here?

**Comment by peterschrammel** [comment]
> Note that this also tweaks the remove_returns behavior, which may not have been intended.

~I deliberately changed that because silently ignoring the failure and returning a nil-irept through a code_typet (!) didn't seem appropriate to me. That case can only occur when the symbol table is out-of-sync with the goto-program, which is clearly a programming error elsewhere.~
Since this is changing behaviour, I will put that change into a separate commit.

**Comment by peterschrammel** [comment]
I revise my statement above. The code is actually used in cbmc/symex_coverage. However, I think that the call there can be simplified if `original_return_type` behaved slightly differently. @tautschnig, please check b29b043 whether this follows your intentions.

**Comment by tautschnig** [comment]
> please check b29b043 whether this follows your intentions.

This looks good to me, and certainly improves the clarity of the code.

**Review by chrisr-diffblue — APPROVED** [review]
Looks OK to me.

**Comment by tautschnig** [comment]
I don't think this should have been merged just yet. I have created #2287 to track this.

**Pull request merged**